### ISSUE TITLE
Class specific predictions (init_mesh)

### ISSUE THREAD

**Issue by trevor-taeyeop** [facebookresearch/meshrcnn#74]
Can I ask why the class specific predictions part is not supported?? 
I tried to make a simple version example. 
Could you advice how to extend class specific predictions??

https://github.com/facebookresearch/meshrcnn/blob/cd43a4b3ab1b7de16e2dc02d85fc6a45f85493a6/meshrcnn/modeling/roi_heads/roi_heads.py#L290

```
    if self.cls_agnostic_voxel:
        with torch.no_grad():
            vox_in = voxel_logits.sigmoid().squeeze(1)  # (N, V, V, V)
            init_mesh = cubify(vox_in, self.cubify_thresh)  # 1
    else:
        with torch.no_grad():
            vox_in = voxel_logits.sigmoid().squeeze(1)  # (N, V, V, V)
            init_meshes = []
            for idx in range(vox_in.shape[0]):
                init_meshes.append(cubify(vox_in[idx:idx+1], self.cubify_thresh))
```

**Comment by gkioxari** [comment]
Our paper makes class agnostic predictions. Early experiments showed that class specific predictions were bad likely because Pix3D has such imbalance of data that it led to some categories having very few positives and thus models to be bad for these classes. On the contrary, class agnostic predictions have the benefit of sharing across across classes, which is crucial for small datasets like Pix3D. 

I added the flag there so that we could extend it for future work. You can easily extend the model to make class specific predictions

* First, your voxel head needs to output`NxCxVxVxV` voxels where `C` is the number of classes. That is already taken care of in the code and you don't need to do anything

https://github.com/facebookresearch/meshrcnn/blob/f61d2c7e6abaa962f5b601739d1968fa5d546f10/meshrcnn/modeling/roi_heads/voxel_head.py#L146

* Then you need to make sure the loss knows what to do in that case. This is again taken care of and you don't need to do anything

https://github.com/facebookresearch/meshrcnn/blob/f61d2c7e6abaa962f5b601739d1968fa5d546f10/meshrcnn/modeling/roi_heads/voxel_head.py#L65-L70

* Finally, you need to make sure that the voxels being fed into the mesh refinement branch in the case of class specific predictions are selected. That is the part that is missing in the codebase (and where you get the ValueError) because it know becomes a modeling choice. If you want to keep the mesh refinement branch class agnostic (as is now the case), you want to select the voxels based on the ground truth class (exactly how they are selected when computing the loss, see above). The easiest way to do this is to edit the [`voxel_rcnn_loss`](https://github.com/facebookresearch/meshrcnn/blob/f61d2c7e6abaa962f5b601739d1968fa5d546f10/meshrcnn/modeling/roi_heads/voxel_head.py#L15) to return the class selected voxels, namely return this

https://github.com/facebookresearch/meshrcnn/blob/f61d2c7e6abaa962f5b601739d1968fa5d546f10/meshrcnn/modeling/roi_heads/voxel_head.py#L70

`pred_voxel_logits` have shape `NxVxVxV` and can now be fed into `cubify`

The code will look something like this:

```
if self.voxel_on:
    voxel_features = self.voxel_pooler(features, proposal_boxes)
    voxel_logits = self.voxel_head(voxel_features)
    loss_voxel, target_voxels, pred_voxel_logits = voxel_rcnn_loss(
        voxel_logits, proposals, loss_weight=self.voxel_loss_weight
    )
    losses.update({"loss_voxel": loss_voxel})
    if self._vis:
        self._misc["target_voxels"] = target_voxels
    if self.cls_agnostic_voxel:
        with torch.no_grad():
            vox_in = voxel_logits.sigmoid().squeeze(1)  # (N, V, V, V)
            init_mesh = cubify(vox_in, self.cubify_thresh)  # 1
    else:
        with torch.no_grad():
            vox_in = pred_voxel_logits.sigmoid()  # (N, V, V, V)
            init_mesh = cubify(vox_in, self.cubify_thresh)  # 1
```

I haven't tested this fyi but you get the picture hopefully!

P.S. My answers are so long :'(

**Comment by trevor-taeyeop** [comment]
Thanks for detailed explanation.

I have an addition question. 

As I understood, different class voxel information makes bad effect of the mesh quality. (ex, If I have two categories (chair and mug cup), chair voxel information is unrelated to mug mesh, so that the quality of mesh could be degraded.)

Q1. In the class specific voxel setup, 
Should I have to extend the mesh refinement for class specific, instead of class agnostic?? Can I ask for your opinion??

**Comment by gkioxari** [comment]
Class agnostic predictions don't have to degrade if you have a good balance of data in your training. Deep neural networks are generally great at projecting objects in higher dimensional feature spaces where different categories live far from each other. Your performance can degrade though if you don't have enough data or the data is not balanced across classes. 

The mesh refinement branch is not setup for making class specific predictions. I believe it's very easy to extend it though.

**Comment by trevor-taeyeop** [comment]
Thank you for a detailed description.
